lib/src/reduction_tree.dart

```diff
@@ -9,136 +9,159 @@
+      : super(
+            definitionName: 'ReductionTreeNode_R${radix}_L${sequence.length}') {
     if (sequence.isEmpty) {
       throw RohdHclException("Don't use ReductionTree "
           'with an empty sequence');
     }
-    sequence = [
+    _sequence = [
       for (var i = 0; i < sequence.length; i++)
         addInput('seq$i', sequence[i], width: sequence[i].width)
     ];
-    this.clk = (clk != null) ? addInput('clk', clk) : null;
-    this.enable = (enable != null) ? addInput('enable', enable) : null;
-    this.reset = (reset != null) ? addInput('reset', reset) : null;
+    _clk = (clk != null) ? addInput('clk', clk) : null;
+    _enable = (enable != null) ? addInput('enable', enable) : null;
+    _reset = (reset != null) ? addInput('reset', reset) : null;
 
-    _computed = reductionTreeRecurse(sequence);
-    addOutput('out', width: _computed.value.width) <= _computed.value;
+    _buildLogic();
   }
 
-  /// Local conditional flop using module reset/enable
-  Logic localFlop(Logic d, {bool doFlop = false}) =>
-      condFlop(doFlop ? clk : null, reset: reset, en: enable, d);
-
-  /// Recursively construct the computation tree
-  ({Logic value, int depth, int flopDepth}) reductionTreeRecurse(
-      List<Logic> seq) {
-    if (seq.length < radix) {
-      return (value: operation(seq), depth: 0, flopDepth: 0);
+  /// Compute the next splitting point taking into account trying to get
+  /// nice powers of the radix as a potential split.
+  static int _nextSplitDistance(
+      int seqRemaining, int radix, int branchesRemaining) {
+    if (branchesRemaining == 1) {
+      return seqRemaining;
+    }
+    final power = (log(seqRemaining) / log(radix)).toInt();
+
+    final chunk = pow(radix, power).toInt();
+
+    if (seqRemaining - chunk - (branchesRemaining - 1) >= 0) {
+      return pow(radix, power).toInt();
+    }
+    return pow(radix, max(power - 1, 0)).toInt();
+  }
+
+  /// Build out the recursive tree
+  void _buildLogic() {
+    if (_sequence.length <= radix) {
+      final value = _operation(_sequence);
+      addOutput('out', width: value.width) <= value;
+      _computed = (value: output('out'), depth: 0, flopDepth: 0);
     } else {
       final results = <({Logic value, int depth, int flopDepth})>[];
-      final segment = seq.length ~/ radix;
       var pos = 0;
       for (var i = 0; i < radix; i++) {
-        final c = reductionTreeRecurse(seq
-            .getRange(pos, (i < radix - 1) ? pos + segment : seq.length)
-            .toList());
-        results.add(c);
-        pos += segment;
+        final end1 =
+            _nextSplitDistance(_sequence.length - pos, radix, radix - i) + pos;
+        final tree = ReductionTree(
+            _sequence.getRange(pos, end1).toList(), _operation,
+            radix: radix,
+            signExtend: signExtend,
+            depthToFlop: depthToFlop,
+            clk: _clk,
+            enable: _enable,
+            reset: _reset);
```

Comment:
do you want to pass `name` down the hierarchy? maybe a modified version of it based on level or something?

Reply:
I will just pass the top level name down.  I don't want to modify that name (say with depth) as that causes Nodes to be uniquified and results in multiple Cells that are basically the same but different because of a name inside.  Because of splitting, we can reuse a Cell Master at different levels of hierarchy when the tree is imbalanced.